controllers/topology/internal/extensions/patches/inline/json_patch_generator.go

```diff
@@ -0,0 +1,372 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package inline implements the inline JSON patch generator.
+package inline
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+	"text/template"
+
+	"github.com/pkg/errors"
+	"github.com/valyala/fastjson"
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/cluster-api/controllers/topology/internal/extensions/patches/api"
+	patchvariables "sigs.k8s.io/cluster-api/controllers/topology/internal/extensions/patches/variables"
+	"sigs.k8s.io/yaml"
+)
+
+// JSONPatchGenerator generates JSON patches for a GenerateRequest based on a ClusterClassPatch.
+type JSONPatchGenerator struct {
+	patch *clusterv1.ClusterClassPatch
+}
+
+// NewJSONPatchGenerator returns a JSONPatchGenerator from a given ClusterClassPatch object.
+func NewJSONPatchGenerator(patch *clusterv1.ClusterClassPatch) api.Generator {
+	return &JSONPatchGenerator{
+		patch: patch,
+	}
+}
+
+// Generate generates JSON patches for the given GenerateRequest based on a ClusterClassPatch.
+func (j *JSONPatchGenerator) Generate(_ context.Context, req *api.GenerateRequest) (*api.GenerateResponse, error) {
+	resp := &api.GenerateResponse{}
+
+	// Loop over all templates.
+	for _, template := range req.Items {
+		// Calculate the list of patches which match the current template.
+		matchingPatches := []clusterv1.PatchDefinition{}
+		for _, patch := range j.patch.Definitions {
+			// Add the patch to the list, if it matches the template.
+			if templateMatchesSelector(&template.TemplateRef, patch.Selector) {
+				matchingPatches = append(matchingPatches, patch)
+			}
+		}
+
+		// Continue if there are no matching patches.
+		if len(matchingPatches) == 0 {
+			continue
+		}
+
+		// Merge template-specific and global variables.
+		// NOTE: To optimize performance, we only do this once for every template.
+		variables, err := mergeVariableMaps(req.Variables, template.Variables)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to merge global and template-specific variables")
+		}
+
+		// Loop over all PatchDefinitions.
+		for _, patch := range matchingPatches {
+			// Generate JSON patches.
+			jsonPatches, err := generateJSONPatches(patch.JSONPatches, variables)
+			if err != nil {
+				return nil, err
+			}
+
+			// Add jsonPatches to the response.
+			resp.Items = append(resp.Items, api.GenerateResponsePatch{
+				TemplateRef: template.TemplateRef,
+				Patch:       *jsonPatches,
+				PatchType:   api.JSONPatchType,
+			})
+		}
+	}
+
+	return resp, nil
+}
+
+// templateMatchesSelector returns true if the template matches the selector.
+func templateMatchesSelector(templateRef *api.TemplateRef, selector clusterv1.PatchSelector) bool {
+	// Check if the apiVersion and kind are matching.
+	if templateRef.APIVersion != selector.APIVersion {
+		return false
+	}
```

Comment:
Thought for future us: Technically we could rely on conversion webhooks here if the `version` part doesn't match.

Reply:
We could, but the main reason we decided to do it this way was to force users to be very explicit about which version of a template they want to patch.

But yeah, we might want to reconsider this in the future.